src/main/java/org/opensearch/knn/training/VectorReader.java

```diff
@@ -0,0 +1,162 @@
+            throw validationException;
+        }
+
+        MappedFieldType fieldType = indicesService.indexServiceSafe(indexMetadata.getIndex()).mapperService()
+                .fieldType(fieldName);
+
+        if (!(fieldType instanceof KNNVectorFieldMapper.KNNVectorFieldType)) {
+            validationException = validationException == null ? new ValidationException() : validationException;
+            validationException.addValidationError("field \"" + fieldName + "\" must be of type KNNVectorFieldType");
+        }
+
+        if (validationException != null) {
+            throw validationException;
+        }
+
+        // Start reading vectors from index
+        ActionListener<SearchResponse> vectorReaderListener = getVectorReaderListener(fieldName,
+                maxVectorCount, 0, vectorConsumer, listener);
+
+        createSearchRequestBuilder(indexName, fieldName, Integer.min(maxVectorCount, searchSize))
+                .execute(vectorReaderListener);
+    }
+
+    @SuppressWarnings("unchecked")
+    private ActionListener<SearchResponse> getVectorReaderListener(String fieldName,
+                                                                   final int maxVectorCount,
+                                                                   final int collectedVectorCount,
+                                                                   Consumer<List<Float[]>> vectorConsumer,
+                                                                   ActionListener<SearchResponse> listener) {
+        return ActionListener.wrap(searchResponse -> {
+            // Get the vectors from the search response
+            // Either add the entire set of returned hits, or maxVectorCount - collectedVectorCount hits
+            SearchHit[] hits = searchResponse.getHits().getHits();
+            int vectorsToAdd = Integer.min(maxVectorCount - collectedVectorCount, hits.length);
+            List<Float[]> trainingData = new ArrayList<>();
+
+            for (int i = 0; i < vectorsToAdd; i++) {
+                trainingData.add(((List<Double>) hits[i].getSourceAsMap().get(fieldName)).stream()
+                        .map(Double::floatValue)
+                        .toArray(Float[]::new));
+            }
+
+            final int updatedVectorCount = collectedVectorCount + trainingData.size();
+
+            // Do something with the vectors
+            vectorConsumer.accept(trainingData);
+
+            if (vectorsToAdd <= 0 || updatedVectorCount >= maxVectorCount) {
+                listener.onResponse(searchResponse);
+            } else {
+                ActionListener<SearchResponse> vectorReaderListener = getVectorReaderListener(fieldName, maxVectorCount,
+                        updatedVectorCount, vectorConsumer, listener);
+
+                // Create a new search that starts where the last search left off
+                createSearchScrollRequestBuilder(searchResponse).execute(vectorReaderListener);
```

Comment:
Can we reuse a scroll request by setting scroll id of the same request?

Reply:
I believe it can be used. Will require some refactoring. I will try to do this now.

Reply:
I reused it with composite aggregation: https://github.com/kaituo/anomaly-detection-1/blob/multiCat4/src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java#L307-L395

Reply:
I see. That makes sense.